packages/dd-trace/src/instrumenter.js

```diff
@@ -112,33 +129,38 @@ class Instrumenter {
           value: true,
           configurable: true
         })
-      })
-    })
 
-    shimmer.massWrap.call(this, nodules, names, function (original, name) {
-      const wrapped = wrapper(original, name)
-      const props = Object.getOwnPropertyDescriptors(original)
-      const keys = Reflect.ownKeys(props)
+        if (Reflect.ownKeys(nodule[name]).includes('_datadog_wrapped')) {
+          nodule[name]._datadog_wrapped = wrapper(nodule[name]._datadog_wrapped)
+          return
+        }
+
+        shimmer.wrap.call(this, nodule, name, function (original, name) {
+          const wrapped = wrapper(original, name)
+          const props = Object.getOwnPropertyDescriptors(original)
+          const keys = Reflect.ownKeys(props)
 
-      // https://github.com/othiym23/shimmer/issues/19
-      for (const key of keys) {
-        if (typeof key !== 'symbol' || wrapped.hasOwnProperty(key)) continue
+          // https://github.com/othiym23/shimmer/issues/19
+          for (const key of keys) {
+            if (typeof key !== 'symbol' || wrapped.hasOwnProperty(key)) continue
 
-        Object.defineProperty(wrapped, key, props[key])
-      }
+            Object.defineProperty(wrapped, key, props[key])
+          }
 
-      return wrapped
+          return wrapped
+        })
+      })
     })
   }
 
   unwrap (nodules, names, wrapper) {
     nodules = [].concat(nodules)
     names = [].concat(names)
 
-    shimmer.massUnwrap.call(this, nodules, names, wrapper)
-
     nodules.forEach(nodule => {
       names.forEach(name => {
+        nodule[name] && delete nodule[name]._datadog_wrapped
```

Comment:
My understanding was that only the internal wrapper would be removed but the prepatch would stay.

Reply:
That's correct. `_datadog_wrapped` is set to the internal wrapper, so the prepatched `nodule[name]` reverts to calling the original function, still preserved in the function scope.

Deleting the property entirely will cause any further calls to `patch` to simply to the same wrapping as it normally would, wrapping the prepatched function, without the property set, so it reverts to behaving like the original.

A more efficient implementation might set the value to null rather than deleting it, so that re-wrapping becomes just as efficient as it was the first time around, but this case doesn't seem particularly likely to me. I can change it anyway.